src/cliargs/core.lua

```diff
@@ -0,0 +1,623 @@
+    end
+
+    local full_msg = cli.name .. ": error: " .. msg .. '; re-run with --help for usage.'
+
+    if not silent then
+      cli.printer.print(full_msg)
+    end
+
+    return nil, full_msg
+  end
+
+  -- Used internally to add an option
+  local function really_add_option(k, expanded_key, v, label, desc, default, callback)
+    local flag = (v == nil) -- no value, so it's a flag
+    local negatable = flag and (expanded_key and expanded_key:find('^%[no%-]') ~= nil)
+    local ek = negatable and expanded_key:sub(6) or expanded_key
+
+    -- guard against duplicates
+    if lookup(k, ek, optional) then
+      error("Duplicate option: " .. (k or ek) .. ", please rename one of them.")
+    end
+
+    if negatable and lookup(nil, "no-"..ek, optional) then
+      error("Duplicate option: " .. ("no-"..ek) .. ", please rename one of them.")
+    end
+
+    -- below description of full entry record, nils included for reference
+    local entry = {
+      key = k,
+      expanded_key = ek,
+      desc = desc,
+      default = default,
+      label = label,
+      flag = flag,
+      negatable = negatable,
+      callback = callback,
+      __display_key = ek or k
+    }
+
+    table.insert(optional, entry)
+  end
+
+  local function get_initial_values()
+    local initial_values = {}
+
+    for _, entry in ipairs(required) do
+      initial_values[entry] = shallow_copy(entry.default)
+    end
+
+    for _, entry in ipairs(optional) do
+      initial_values[entry] = shallow_copy(entry.default)
+    end
+
+    if optargument.key then
+      if optargument.maxcount > 1 then
+        initial_values[optargument] = {}
+      else
+        initial_values[optargument] = optargument.default
+      end
+    end
+
+    return initial_values
+  end
+
+  -- ------------------------------------------------------------------------ --
+  -- PUBLIC API
+  -- ------------------------------------------------------------------------ --
+
+  --- CONFIG
+
+  --- Assigns the name of the program which will be used for logging.
+  function cli:set_name(in_name)
+    cli.name = in_name
+  end
+
+  --- Write down a brief, 1-liner description of what the program does.
+  function cli:set_description(in_description)
+    cli.description = in_description
+  end
+
+  --- Sets the amount of space allocated to the argument keys and descriptions
+  --- in the help listing.
+  ---
+  --- The sizes are used for wrapping long argument keys and descriptions.
+  ---
+  --- @param {number} [key_cols=0]
+  ---        The number of columns assigned to the argument keys, set to 0 to
+  ---        auto detect.
+  ---
+  --- @param {number} [desc_cols=0]
+  ---        The number of columns assigned to the argument descriptions, set to
+  ---        0 to auto set the total width to 72.
+  function cli:set_colsz(key_cols, desc_cols)
+    colsz = { key_cols or colsz[1], desc_cols or colsz[2] }
+  end
+
+  function cli:set_silent(in_silent)
+    silent = in_silent
+  end
```

Comment:
Probably need a `cli:get_silent()` or `cli:is_silent()` public function so that a custom error handler can be silent if `silent` is true